File: it.stefano.bls.domain/lib/src/main/java/led/Led.java
package led;

import interf.ILed;

public abstract class Led implements ILed {

	private boolean state = false;
	
	@Override
	public void on() {
		setState(true);
		doOn();
	}

	protected abstract void doOn();

	private void setState(boolean aState) {
		this.state = aState;
	}

	@Override
	public void off() {
		setState(true);
		doOff();
	}

	protected abstract void doOff();

	@Override
	public boolean isOn() {
		return state;
	}

}

File: it.stefano.bls.arduino/libraries/Led/Led.cpp
#include "Led.h"
#include "LedDevice.h"

Led::Led(LedDevice* device) {
  this->device = device;
  this->off();
}
void Led::on() {
  this->state = true;
  this->device->on();
}
void Led::off() {
  this->state = false;
  device->off();
}
bool Led::isOn() {
  return state;
}
